### PR TITLE
Implementation of #82 - Integrate Data into Graphs

### PR DESCRIPTION
**Pull request by SaiVarunVaranasi**
For reference: The older PRs of this Dev branch are:

#98 #108 

And the related feature is #82

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -53,8 +56,9 @@
     "devDependencies": {
         "@testing-library/jest-dom": "^5.12.0",
         "@testing-library/react": "^11.2.7",
+        "axios-mock-adapter": "^1.19.0",
         "gh-pages": "^2.2.0",
-        "jest": "26.6.0",
+        "jest": "^26.6.0",
```
Comment (Waldleufer) [inline frontend/package.json:61]:
```suggestion
        "jest": "26.6.0",
```

I believe this is breaking the frontend pipeline, since the testing requires an exact version. Why did you change this @SaiVarunVaranasi and can we change it back?

File: frontend/package.json [hunk 2]
```diff
@@ -8,7 +8,10 @@
         "apexcharts": "^3.26.1",
         "axios": "^0.21.1",
         "cors": "^2.8.5",
+        "enzyme": "^3.11.0",
+        "enzyme-adapter-react-16": "^1.15.6",
         "html2canvas": "^1.0.0-rc.7",
+        "jest-cli": "^20.0.4",
```
Comment (Waldleufer) [inline frontend/package.json:14]: Don't forget to update the BOM in Google Drive if you add new dependencies here @SaiVarunVaranasi

**Review by Waldleufer — COMMENTED** [review]
Reviewed Package.json, got some questions

File: frontend/test/setupTests.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import { configure } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+configure({ adapter: new Adapter(), disableLifecycleMethods: true });
```
Comment (Waldleufer) [inline frontend/test/setupTests.js:3]: What does this do?

File: frontend/test/tempPolyfills.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+const requestAnimationFrame = (global.requestAnimationFrame = (callback) => {
+    setTimeout(callback, 0);
+});
+export default requestAnimationFrame;
```
Comment (Waldleufer) [inline frontend/test/tempPolyfills.js:4]: Please add some documentation what this is meant to do ☺

File: frontend/test/setupTests.js (hunk 3 above)
Comment (Waldleufer) [reply frontend/test/setupTests.js:3]: (Maybe you can add one line of documentation here)

**Review by Waldleufer — COMMENTED** [review]
:)

File: frontend/src/components/details/TableComponent.jsx [hunk 5]
```diff
@@ -16,66 +19,18 @@ class TableComponent extends Component {
...
-            {
-                key: 'row-6',
-                impactCategory: 'Nonrenewable energy',
-                unit: 'MJ eq',
-                total: '2,781,500,619',
-                materialsLPT: '14,672,424',
-                manufacturing: '1,454,845',
-                operations: '2,774,610,300',
-                endOfLife: '–9,236,950'
+                unit: getImpactAssessmentData[6],
```
Comment (ScoutAtlas) [inline frontend/src/components/details/TableComponent.jsx:22]: The order of the ImpactAssessmentData is set by the SimaPro Data?

File: frontend/src/components/login/LoginComponent.js [hunk 6]
```diff
@@ -15,19 +14,12 @@ function LoginComponent() {
     const [, setState] = useContext(GlobalContext);
 
     return (
-        <div
-            className='flex-container'
-            style={{ backgroundColor: uniformStyle.color.secondaryBackgroundColor }}
-        >
+        <div className='flex-container'>
```
Comment (ScoutAtlas) [inline frontend/src/components/login/LoginComponent.js:17]: Why are we changing styling in this branch?

File: frontend/src/components/login/loginstyle.css [hunk 7]
```diff
@@ -11,17 +8,23 @@
     align-items: center;
     height: 100%;
     overflow: auto;
+    color: 'white';
+    background: radial-gradient(circle at top, #000000, #1b1534, #4d217a, #643082);
```
Comment (ScoutAtlas) [inline frontend/src/components/login/loginstyle.css:12]: Same here: why are we changing styling/colors?
Isn't there an extra branch for that?

File: frontend/src/components/productGrid/ProductGridComponent.js [hunk 8]
```diff
@@ -51,6 +51,9 @@ function ProductGridComponent({ selectedCategory }) {
     const classes = useStyles();
 
     if (productList === [] || productList === undefined || productList === null) {
+        console.error(
```
Comment (ScoutAtlas) [inline frontend/src/components/productGrid/ProductGridComponent.js:54]: smart :) 👍

File: frontend/src/interface/projectInterface.js [hunk 9]
```diff
@@ -29,44 +35,98 @@ export function getModels() {
...
+    materialCompositionLabels = Array.from(compositionLabels);
+
+    console.log(materialCompositionLabels);
+}
+
+/**
+ * Getter method to recieve the filtered Material Composititon Labels from API
+ */
+export async function getMaterialCompositionLabels() {
+    return materialCompositionLabels;
```
Comment (ScoutAtlas) [inline frontend/src/interface/projectInterface.js:82]: Have we implemented the color mapping for the materials yet?

File: frontend/src/interface/projectInterface.js [hunk 10]
```diff
@@ -29,44 +35,98 @@ export function getModels() {
...
+export function setLifeCycleStages(assessmentModels) {
+    lifeCycleStages = Array.from(assessmentModels);
+    console.log('Life Cycle Stages');
+    console.log(lifeCycleStages);
+}
+/**
+ * Getter method to recieve the filtered Life Cycle Stages from API
+ */
+export function getLifeCycleStages() {
+    return lifeCycleStages;
```
Comment (ScoutAtlas) [inline frontend/src/interface/projectInterface.js:111]: Aren't the life cycle stages fixed?
Do we want/need to get them every time?

**Review by ScoutAtlas — COMMENTED** [review]
Implementation looks good in general.
Just some questions regarding, why we are doing style changes in this branch.
And some data handling questions.

File: frontend/src/interface/projectInterface.js (hunk 9 above)
Comment (Waldleufer) [reply frontend/src/interface/projectInterface.js:82]: I don't think so

File: frontend/test/setupTests.js (hunk 3 above)
Comment (SaiVarunVaranasi) [reply frontend/test/setupTests.js]: I removed these as the testing is not implemented properly.

File: frontend/src/components/login/loginstyle.css (hunk 7 above)
Comment (SaiVarunVaranasi) [reply frontend/src/components/login/loginstyle.css:12]: I think these changes are shown as we tried to resolve the merge conflicts

File: frontend/src/interface/projectInterface.js (hunk 9 above)
Comment (SaiVarunVaranasi) [reply frontend/src/interface/projectInterface.js:81]: This was updated yesterday and the changes were reverted. As we do not want to receive the label data from API

**Review by ScoutAtlas — APPROVED** [review]

**Comment by Waldleufer** [comment]
@SaiVarunVaranasi @ScoutAtlas How can I test the implementations? only with VA71199 ? Or with another product as well?

File: frontend/src/components/details/TableComponent.jsx [hunk 11]
```diff
@@ -3,16 +3,15 @@ import React, { Component } from 'react';
...
-
 /**
  * displays the impact catagories table of the selected model of the related product.
  */
 class TableComponent extends Component {
     state = {
-        headers: getImpactCategoriesTableHeaders(this.props.modelId),
-        rows: getImpactCategoriesTableData(this.props.modelId)
+        headers: getImpactCategoriesTableHeaders(),
+        rows: getImpactAssessmentData()
```
Comment (Waldleufer) [inline frontend/src/components/details/TableComponent.jsx:14]: how do these know, what data to fetch without passing any data to them here?

File: frontend/src/components/productGrid/ProductGridComponent.js [hunk 12]
```diff
@@ -48,18 +45,21 @@ function ProductGridComponent({ selectedCategory }) {
     ];
 
     if (productList === [] || productList === undefined || productList === null) {
+        console.error(
+            'Products data not recieved.Please verify the API calls and Backend Connection.'
+        );
         return <LoadingComponent />;
```
Comment (Waldleufer) [inline frontend/src/components/productGrid/ProductGridComponent.js:51]:
```suggestion
        return <LoadingComponent loading />;
```

otherwise it will hide itself 😉

File: frontend/src/interface/BackendConnect.js [hunk 13]
```diff
@@ -1,58 +1,113 @@
...
+            headers
+        },
+        { body }
+    );
     // .then(function (data) {
     //     const items = data;
     //     calcId = items.data.Result.CalculationId;
     // });
-    console.log('RESULT!!!!!!1');
+    console.log('RESULT!!!!!!2');
```
Comment (Waldleufer) [inline frontend/src/interface/BackendConnect.js:96]:
```suggestion
    console.log('RESULT of the SimaPro/api/calculation/${projectId} call ');
```

File: frontend/src/interface/processBackendData.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+import {
+    setMaterialCompositionLabels,
+    setMaterialCompositionData,
+    setImpactAssessmentData,
+    setColumnChartData
+} from 'interface/projectInterface';
+
```
Comment (Waldleufer) [inline frontend/src/interface/processBackendData.js:7]: Please edit the callback description before commiting this requested change

```suggestion

/**
 * Filters the carbon impact data recieved from API.
 * Filter the Carbon Values of GlobalWarming
 * Maps the Carbon Values and its corresponding life cycle stage.
 * @param data data recieved from PostCalculationRequest
 * @param callback TODO: PLEASE EDIT ME BEFORE COMMITTING!
 */
```

File: frontend/src/interface/processBackendData.js [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        materialMap.set(finalMaterials[i][0], finalMaterials[i][6]);
+    }
+
+    /**
+     * Filters the carbon impact data recieved from API.
+     * Filter the Carbon Values of GlobalWarming
+     * Maps the Carbon Values and its corresponding life cycle stage.
+     * @param data data recieved from PostCalculationRequest
+     */
+
```
Comment (Waldleufer) [inline frontend/src/interface/processBackendData.js:30]:
```suggestion
```

**Review by Waldleufer — COMMENTED** [review]
There are still lots of console outputs that we will need to clean up at some point in the future.

Looking good else 🙂

File: .github/workflows/Reactjs.yml [hunk 16]
```diff
@@ -5,14 +5,17 @@ name: Frontend CI
 
 on:
   push:
-    branches: [ main, release ]
+    branches: [ main, release, dev-#82 ]
```
[the remark this answers is not in the document]
Comment (SaiVarunVaranasi) [reply .github/workflows/Reactjs.yml:8]: Yeah... I added this in order to fix the CI pipeline for the branch.

File: frontend/jest.config.js [hunk 17]
```diff
@@ -187,3 +187,9 @@ module.exports = {
     // Whether to use watchman for file crawling
     // watchman: true,
 };
+module.exports = {
+    moduleNameMapper: {
+        '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga|css|png|less|scss)$':
+            'identity-obj-proxy'
+    }
+};
```
[the remark this answers is not in the document]
Comment (SaiVarunVaranasi) [reply frontend/jest.config.js:195]: This is used for automated testing. Which is now dropped.I will check if it needed and make the necessary changes.

File: frontend/src/components/details/ColumnChartComponent.jsx [hunk 18]
```diff
@@ -1,18 +1,18 @@
...
  * Column Chart
  *
- * @author Julian Oelhaf
  */
 const ColumnChartComponent = () => {
+    const values = getColumnChartData();
     const series = [
         {
             name: 'Global warming in kg CO2 equivalents',
             // TODO: this data needs to be recieved from backend
```
[the remark this answers is not in the document]
Comment (SaiVarunVaranasi) [reply frontend/src/components/details/ColumnChartComponent.jsx:14]: Yes. This is done.

File: frontend/src/components/details/TableComponent.jsx (hunk 11 above)
Comment (SaiVarunVaranasi) [reply frontend/src/components/details/TableComponent.jsx:14]: For now, we are only getting the calculation result based on the Project Id and not the model Id. We need to implement the calculation call based on model Id at a later point in time. Also, the calculation data with respect to model Id will also be called from BackendConnect.js processed at ProjectInterface.js and processBackendData.js. So don't think we need to pass any value here...we only send one API request and process the calculation data for filling the details page.

**Review by SaiVarunVaranasi — COMMENTED** [review]
Looks good for me. @Waldleufer I made some changes as suggested by you please reverify them.

**Comment by SaiVarunVaranasi** [comment]
> @SaiVarunVaranasi @ScoutAtlas How can I test the implementations? only with VA71199 ? Or with another product as well?
@Waldleufer @ScoutAtlas 
You can test the details page for the below products:
Reactor 1#ID:Large Reactor#Scenario:copper_baseline
Service ABC#ID:Large ABC Service#Scenario:copper_scenario2

File: frontend/src/interface/processBackendData.js (hunk 14 above)
Comment (ScoutAtlas) [reply frontend/src/interface/processBackendData.js:7]: I added a few lines, do you think that's enough?

**Review by Waldleufer — APPROVED** [review]

File: frontend/src/components/details/ScenarioComponent.js [hunk 19]
```diff
@@ -94,7 +101,7 @@ ScenarioComponent.propTypes = {
     onExportClick: PropTypes.func.isRequired,
```
[the remark this answers is not in the document]
Comment (Waldleufer) [reply frontend/src/components/details/ScenarioComponent.js:94]: for better console feedbacks during testing. It makes sure we use a component consistently.
